### PR TITLE
[move] Relaxed one-time withess verfification rules

### PR DESCRIPTION
**Pull request by awelc**
This resolves https://github.com/MystenLabs/sui/issues/5041. The problem reported there was that types accidentally called the same as the capitalized module name they are defined in are reported as one-time witness candidates (thus preventing successful module compilation) even if the have a different structure than is expected from a one-time witness type.

This PR relaxes the rules that the verifier enforces by allowing a type with the same name as capitalized module name to NOT be considered a one-time witness candidate if it has a structure different than one boolean field. This check (whether a struct has one boolean field is now performed in the native function verifying if a value is of one-type witness type.

### CODE REVIEW

File: crates/sui-framework/src/natives/types.rs [hunk 1]
```diff
@@ -20,19 +23,50 @@ pub fn is_one_time_witness(
...
+        Some(MoveTypeLayout::Struct(l)) => l,
+        _ => return Ok(NativeResult::ok(cost, smallvec![Value::bool(false)])),
+    };
+
+    let has_one_bool_field = match struct_layout {
+        MoveStructLayout::Runtime(vec) => {
+            if !vec.len() == 1 {
+                false
+            } else {
+                matches!(vec[0], MoveTypeLayout::Bool)
```
Comment (tnowacki) [inline crates/sui-framework/src/natives/types.rs:42]: You might be able to do `matches!(vec.as_slice(), [MoveTypeLayout::Bool])`

File: crates/sui-verifier-transactional-tests/tests/one_time_witness/wrong_field_type_with_init.move [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+// SPDX-License-Identifier: Apache-2.0
+
+// invalid, wrong struct field type
+
+//# init --addresses test=0x0
+
+//# publish
+module test::m {
+
+    struct M has drop { value: u64 }
```
Comment (tnowacki) [inline crates/sui-verifier-transactional-tests/tests/one_time_witness/wrong_field_type_with_init.move:11]: Maybe just duplicate this test to make sure we can still construct this module

Something like 
```
test::m {
struct M has drop { value: u64 }

fun foo() {
  M { value: 0 };
  M { value: 1 };
}
}

**Review by tnowacki — APPROVED** [review]
Approving to unblock, feel free to update the rust style (or not) and the other test would be nice for clarity, but I don't think it is necessary for coverage.

File: crates/sui-framework/src/natives/types.rs (hunk 1 above)
Comment (awelc) [reply crates/sui-framework/src/natives/types.rs:42]: Rust's cleverness never fails to surprise me! Thanks!

File: crates/sui-verifier-transactional-tests/tests/one_time_witness/wrong_field_type_with_init.move (hunk 2 above)
Comment (awelc) [reply crates/sui-verifier-transactional-tests/tests/one_time_witness/wrong_field_type_with_init.move:11]: I kind of already did (in `wrong_field_type.move` above but admittedly without explicit instantiation which I now added.

File: crates/sui-framework/src/natives/types.rs (hunk 1 above)
Comment (tnowacki) [reply crates/sui-framework/src/natives/types.rs]: The slice pattern matches are relatively new :) So probably won't see them around too too much

**Pull request merged**